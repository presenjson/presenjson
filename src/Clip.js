import React from 'react';
import PropTypes from 'prop-types';

const Clip = (props) => {
    return (<div className='clip'>
        {props.children}
    </div>);
}

Clip.propTypes = {
    length: PropTypes.number
};

Clip.defaultProps = {
    length: 0
}

export default Clip;
