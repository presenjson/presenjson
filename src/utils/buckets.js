import React from 'react';
import ReactDOMServer from 'react-dom/server';
import * as R from 'ramda';

const [POSITION, LENGTH] = [0, 1];
const fullLength = R.pipe(R.map((o) => o.props.length + o.props.delay), R.sum);

export const getLengthsAndPositions = async (components, data) => {
    let pos = 0;
    const position = (delay, length) => {
        const clipStarts = pos + delay;
        pos += length;
        return clipStarts;
    };
    const lengths = await Promise.all(components.map(getLength(data)));

    return components.map((c, i) => ({
        delay: c.props.delay,
        length: lengths[i],
        fullLength: lengths[i] + c.props.delay,
        position: position(c.props.delay, lengths[i])
    }));
};

export const getLength = R.curry((data, child) => {
    return new Promise((resolve, reject) => {
        if (child.props.length) {
            return resolve(child.props.length);
        }

        const Component = child.props.component;
        const onLoad = (length) => resolve(length);

        ReactDOMServer.renderToString(
            <Component {...child.props} data={data} onLoad={onLoad} />
        );
    });
});

export default (children, data) => {
    return new Promise(async (resolve, reject) => {
        const allLengths = await getLengthsAndPositions(children, data);
        const length = allLengths.reduce((a, b) => a + b.fullLength, 0);

        const numOfBuckets = Math.ceil(length / 1000) + 1;
        const bucks = Array(numOfBuckets).fill([], 0, numOfBuckets);

        allLengths.forEach((c, i) => {
            const { length, position } = c;

            if (position % 1000 + length >= 1000) {
                const lastBucket = Math.floor((position + length) / 1000);
                bucks[lastBucket] = [...bucks[lastBucket], i];
            }

            for (let k = position; k < length + position; k += 1000) {
                const b = Math.floor(k / 1000);
                bucks[b] = [...bucks[b], i];
            }
        });

        const buckets = R.map(R.uniq, bucks); // duplicate entries?

        return resolve({
            buckets,
            length,
            positions: R.pluck('position', allLengths),
            lengths: R.pluck('length', allLengths)
        });
    });
};
