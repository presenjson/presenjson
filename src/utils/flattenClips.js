import * as R from 'ramda';
import { Group } from '../ClipGroup';

export default R.pipe(
    R.of,
    R.flatten,
    R.map(R.when((x) => x.type.name === 'ClipGroup', (x) => Group(x.props))),
    R.flatten
);
