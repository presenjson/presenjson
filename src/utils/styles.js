export const blur = (amount = 3) => ({
    filter: `blur(${amount}px)`
});
