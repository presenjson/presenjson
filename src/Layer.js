import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import cx from 'classnames';
import Screen from './Screen';
import buckets from './utils/buckets';
import flattenClips from './utils/flattenClips';

export default class Layer extends Component {
    constructor(...args) {
        super(...args);
        this.time = this.props.startAt;
        this.mountedAt = 0;
        this.state = {
            bucket: Math.ceil(this.time / 1000) - 1,
            onScreen: [],
            buckets: []
        };

        this.loadClips();
    }

    async loadClips() {
        this.clips = flattenClips(this.props.children);
        const state = await buckets(this.clips, this.props.data);
        this.setState(state);
        this.props.onLoad(state.length);
    }

    componentDidMount() {
        this.mountedAt = Date.now();

        requestAnimationFrame(this.loop);
    }

    isOnScreen = (o) =>
        this.state.positions[o] <= this.time &&
        this.state.positions[o] + this.state.lengths[o] >= this.time;

    loop = () => {
        const current = Date.now();

        if (!this.props.paused) {
            this.time = current - this.mountedAt;
            const bucket = Math.ceil(this.time / 1000) - 1;
            const activeBucket = this.state.buckets[bucket] || [];
            const onScreen = activeBucket.filter(this.isOnScreen);

            this.setState({ bucket, onScreen: onScreen[0] });
        } else {
            this.mountedAt = current - this.time;
        }

        requestAnimationFrame(this.loop);
    };

    render() {
        const currentBucket = this.state.buckets[this.state.bucket] || [];
        const classNames = cx({
            layer: true,
            ...R.pick(['background', 'foreground'], this.props)
        });

        return (
            <div className={classNames}>
                {currentBucket.map((i) => (
                    <Screen
                        {...this.clips[i].props}
                        data={this.props.data}
                        paused={this.props.paused}
                        onScreen={this.state.onScreen === i}
                        key={`${this.props.layer}-screen-${i}`}
                        i={i}
                    />
                ))}
            </div>
        );
    }
}

Layer.defaultProps = {
    children: [],
    startAt: 0
};

Layer.propTypes = {
    layer: PropTypes.number,
    startAt: PropTypes.number,
    paused: PropTypes.bool,
    onLoad: PropTypes.func,
    data: PropTypes.any,
    children: PropTypes.node
};
