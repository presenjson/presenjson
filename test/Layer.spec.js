import React from 'react';
import { shallow } from 'enzyme';
import Layer from '../src/Layer';
import Clip from '../src/Clip';
import * as buckets from '../src/utils/buckets';
import * as flattenClips from '../src/utils/flattenClips';

describe('Layer', () => {
    it('constructor sets default state/vars and calls load', () => {
        const loadClips = jest
            .spyOn(Layer.prototype, 'loadClips')
            .mockImplementation(() => {});

        const layer = new Layer({ startAt: 100, children: 'the-children' });

        expect(layer.time).toBe(100);
        expect(layer.mountedAt).toBe(0);
        expect(layer.state).toEqual({ bucket: 0, onScreen: [], buckets: [] });
        expect(loadClips).toHaveBeenCalled();
    });
    describe('load', () => {
        it('calls onLoad with length and sets state', async () => {
            const state = {
                length: 12345,
                other: 'state'
            };
            const setState = jest
                .spyOn(Layer.prototype, 'setState')
                .mockImplementation(() => {});
            const bucketSpy = jest
                .spyOn(buckets, 'default')
                .mockReturnValue(Promise.resolve(state));
            const flattenClipsSpy = jest
                .spyOn(flattenClips, 'default')
                .mockReturnValue('the-clips');

            const onLoad = jest.fn();
            const Component = () => false;
            const children = [<Component length={1000} key={1} />];

            const layer = new Layer({
                startAt: 100,
                children,
                onLoad,
                data: 'the-data'
            });

            await layer.loadClips();

            expect(onLoad).toHaveBeenCalledWith(12345);
            expect(setState).toHaveBeenCalledWith(state);
            expect(bucketSpy).toHaveBeenCalledWith('the-clips', 'the-data');
            expect(flattenClipsSpy).toHaveBeenCalledWith(children);
            expect(layer.clips).toEqual('the-clips');
        });
    });
    describe('componentDidMount', () => {
        it('sets mountedAt and starts animation loop', () => {
            jest.spyOn(Date, 'now').mockReturnValue(123456789);
            jest
                .spyOn(global, 'requestAnimationFrame')
                .mockImplementation(() => {});
            jest
                .spyOn(Layer.prototype, 'loadClips')
                .mockImplementation(() => {});

            const layer = new Layer({});
            layer.loop = jest.fn();
            layer.componentDidMount();

            expect(layer.mountedAt).toBe(123456789);
            expect(global.requestAnimationFrame).toHaveBeenCalledWith(
                layer.loop
            );
        });
    });
    describe('isOnScreen', () => {
        it('checks if clip is currently on being displayed', () => {
            jest
                .spyOn(Layer.prototype, 'loadClips')
                .mockImplementation(() => {});
            const relativeTime = 2000;

            const layer = new Layer({});
            layer.state.lengths = [0, 0, 100, 300];
            layer.state.positions = [0, 0, 1950, 2100];
            layer.time = relativeTime;
            const is2OnScreen = layer.isOnScreen(2);
            const is3OnScreen = layer.isOnScreen(3);

            expect(is2OnScreen).toBeTruthy();
            expect(is3OnScreen).toBeFalsy();
        });
    });
    describe('loop', () => {
        it('calls requestAnimationFrame with itself', () => {
            jest
                .spyOn(Layer.prototype, 'loadClips')
                .mockImplementation(() => {});
            jest
                .spyOn(global, 'requestAnimationFrame')
                .mockImplementation(() => {});

            const layer = new Layer({});
            layer.setState = jest.fn();

            layer.loop();

            expect(requestAnimationFrame).toHaveBeenCalledWith(layer.loop);
        });
        it('sets mountedAt when paused', () => {
            jest.spyOn(Date, 'now').mockReturnValue(1000000);
            jest
                .spyOn(Layer.prototype, 'loadClips')
                .mockImplementation(() => {});
            jest
                .spyOn(global, 'requestAnimationFrame')
                .mockImplementation(() => {});

            const layer = new Layer({ paused: true });
            layer.mountedAt = 22222222;
            layer.time = 2000;

            layer.loop();

            expect(layer.mountedAt).toBe(998000);
        });
        it('sets current bucket and clip on screen', () => {
            jest.spyOn(Date, 'now').mockReturnValue(334444);
            jest
                .spyOn(Layer.prototype, 'loadClips')
                .mockImplementation(() => {});
            jest
                .spyOn(global, 'requestAnimationFrame')
                .mockImplementation(() => {});
            const setState = jest
                .spyOn(Layer.prototype, 'setState')
                .mockImplementation(() => {});

            const layer = new Layer({ paused: false });
            const isOnScreen = jest
                .spyOn(layer, 'isOnScreen')
                .mockImplementation(() => true);
            layer.mountedAt = 333111;
            layer.state.buckets = [[], [3, 4, 5]];

            layer.loop();

            expect(layer.time).toBe(1333);
            expect(setState).toHaveBeenCalledWith({ bucket: 1, onScreen: 3 });
            expect(isOnScreen).toHaveBeenCalledWith(3, 0, [3, 4, 5]);
            expect(isOnScreen).toHaveBeenCalledWith(4, 1, [3, 4, 5]);
            expect(isOnScreen).toHaveBeenCalledWith(5, 2, [3, 4, 5]);
        });
    });
    describe('render', () => {
        it('renders correctly on empty bucket', () => {
            const layer = shallow(<Layer onLoad={() => {}} />);
            expect(layer).toMatchSnapshot();
        });
        it('renders correctly on active bucket', () => {
            const layer = shallow(
                <Layer onLoad={() => {}}>
                    <Clip length={1000} component="the-component" />
                    <Clip length={1000} component="the-component" />
                    <Clip length={500} component="the-component" />
                    <Clip length={500} component="the-component" />
                </Layer>
            );
            layer.setState({
                bucket: 1,
                buckets: [[], [2, 3]]
            });
            expect(layer).toMatchSnapshot();
        });
    });
});
