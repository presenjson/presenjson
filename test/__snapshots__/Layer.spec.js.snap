// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer render renders correctly on active bucket 1`] = `
<div
  className="layer"
>
  <Screen
    component="the-component"
    delay={0}
    i={2}
    key="undefined-screen-2"
    length={500}
    onScreen={false}
  />
  <Screen
    component="the-component"
    delay={0}
    i={3}
    key="undefined-screen-3"
    length={500}
    onScreen={false}
  />
</div>
`;

exports[`Layer render renders correctly on empty bucket 1`] = `
<div
  className="layer"
/>
`;
